File: 3/final.py
def read_file(file):
    with open(file) as f:
        return f.read()
   
import re
   
def part1(txt):
    pattern = r"mul\((-?\d+),\s*(-?\d+)\)"
    results = re.findall(pattern, txt)
    return sum(int(a) * int(b) for a, b in results)
      
def part2(txt):
    parts = txt.split("do()")
    executable_parts = [code.split("don't()")[0] for code in parts]
    return sum(part1(code) for code in executable_parts)


if __name__ == "__main__":
    l = read_file("3/input.txt")
    print("part1 :", part1(l))
    print("part2 :", part2(l))

File: 3/final.ml
let read filename =
  let file = open_in filename in
  let rec read_lines acc =
    try
      let line = input_line file in
      read_lines (line :: acc)
    with End_of_file ->
      close_in file;
      List.rev acc
  in
  String.concat "\n" (read_lines [])

let part1 txt =
  let re = Str.regexp "mul(\\(-?[0-9]+\\),\\s*\\(-?[0-9]+\\))" in
  let rec aux pos acc =
    try
      let pos = Str.search_forward re txt pos in
      let a = int_of_string (Str.matched_group 1 txt) in
      let b = int_of_string (Str.matched_group 2 txt) in
      aux (pos + 1) (acc + a * b)
    with Not_found -> acc
  in
  aux 0 0

let part2 txt =
  let parts = Str.split (Str.regexp "do()") txt in
  let rec aux parts acc =
    match parts with
    | [] -> acc
    | hd :: tl ->
      let s = Str.split (Str.regexp "don't()") hd in
      aux tl (acc + part1 (List.hd s))
  in
  aux parts 0

let () =
  let txt = read "3/input.txt" in
  Printf.printf "Part 1: %d\n" (part1 txt);
  Printf.printf "Part 2: %d\n" (part2 txt)